stateresolution.go

```diff
@@ -0,0 +1,279 @@
+			r.resolvedCreate = event
+		}
+	case "m.room.power_levels":
+		if event.StateKeyEquals("") {
+			r.resolvedPowerLevels = event
+		}
+	case "m.room.join_rules":
+		if event.StateKeyEquals("") {
+			r.resolvedJoinRules = event
+		}
+	case "m.room.member":
+		r.resolvedMembers[*event.StateKey()] = event
+	case "m.room.third_party_invite":
+		r.resolvedThirdPartyInvites[*event.StateKey()] = event
+	default:
+		panic(fmt.Errorf("Unexpected auth event with type %q", event.Type()))
+	}
+}
+
+// Remove the auth event with the given type and state key.
+func (r *stateResolver) removeAuthEvent(eventType, stateKey string) {
+	switch eventType {
+	case "m.room.create":
+		if stateKey == "" {
+			r.resolvedCreate = nil
+		}
+	case "m.room.power_levels":
+		if stateKey == "" {
+			r.resolvedPowerLevels = nil
+		}
+	case "m.room.join_rules":
+		if stateKey == "" {
+			r.resolvedJoinRules = nil
+		}
+	case "m.room.member":
+		r.resolvedMembers[stateKey] = nil
+	case "m.room.third_party_invite":
+		r.resolvedThirdPartyInvites[stateKey] = nil
+	default:
+		panic(fmt.Errorf("Unexpected auth event with type %q", eventType))
+	}
+}
+
+// resolveAndAddAuthBlocks resolves each block of conflicting auth state events in a list of blocks
+// where all the blocks have the same event type.
+// Once every block has been resolved the resulting events are added to the events used for auth checks.
+// This is called once per auth event type and state key pair.
+func (r *stateResolver) resolveAndAddAuthBlocks(blocks [][]Event) {
+	start := len(r.result)
+	for _, block := range blocks {
+		if event := r.resolveAuthBlock(block); event != nil {
+			r.result = append(r.result, *event)
+		}
+	}
+	// Only add the events to the auth events once all of the events with that type have been resolved.
+	// (This is done to avoid the result of state resolution depending on the iteration order)
+	for i := start; i < len(r.result); i++ {
+		r.addAuthEvent(&r.result[i])
+	}
+}
+
+// resolveAuthBlock resolves a block of auth events with the same state key to a single event.
+func (r *stateResolver) resolveAuthBlock(events []Event) *Event {
+	// Sort the events by depth and sha1 of event ID
+	block := make([]conflictedEvent, len(events))
+	for i := range events {
+		event := &events[i]
+		block[i] = conflictedEvent{
+			depth:       event.Depth(),
+			eventIDSHA1: sha1.Sum([]byte(event.EventID())),
+			event:       event,
+		}
+	}
+	sort.Sort(conflictedEventSorter(block))
+
+	// Pick the "oldest" event, that is the one with the lowest depth, as the first candidate.
+	result := block[0].event
+	// Temporarily add the candidate event to the auth events.
+	r.addAuthEvent(result)
+	for i := 1; i < len(block); i++ {
+		event := block[i].event
+		// Check if the next event passes authentication checks against the current candidate.
+		if Allowed(*event, r) == nil {
+			// If the event passes authentication checks pick it as the current candidate.
+			result = event
+			r.addAuthEvent(result)
```

Comment:
Eurgh, this is confusing. Can you please state the rules in a block comment somewhere, otherwise it just makes it look like this has errors in it. From what I can tell:
- if there's only 1 candidate event, choose it (would be nice if we bailed early rather than the subtle `i := 1` skipover)
- else there is >1 candidate event, does it pass auth checks?:
    * YES: It becomes the new candidate. Go to next event.
    * NO: Return the last candidate which passed auth checks.

In all cases, `stateResolver` needs to be "reset" to the previous auth event set. Is this even remotely correct?